Streaming gateway WIP

kalabukdima opened the pull request.

kalabukdima committed:
Implement streaming gateway MVP

kalabukdima committed:
Add prometheus metrics

kalabukdima committed:
Set app level timeouts

Send additional request when all ongoing ones wait for too long

kalabukdima committed:
Add transport metrics

kalabukdima committed:
Add dynamic worker priorities

kalabukdima committed:
Add sliding percentile timeouts

kalabukdima committed:
Update dependencies

- Update to the latest transport crate
- Downgrade rustls to 0.23.10 to avoid cert error
  https://github.com/subsquid/query-gateway/issues/2
- Use subsquid-datasets crate from archive-router repo

kalabukdima committed:
Filter workers by version

kalabukdima committed:
Fetch chunk with multiple requests if it's too big

If a worker has responded with only a subset of blocks
in a chunk, send continuation request for the rest of
the chunk repeatedly until the full chunk is fetched

kalabukdima committed:
Use block range from query

kalabukdima committed:
Update known chunks in background

kalabukdima committed:
Add dockerfile

kalabukdima committed:
Structure source files

kalabukdima committed:
Merge branch 'streaming'

kalabukdima committed:
Track on-chain epochs

kalabukdima committed:
Don't hold mutex across await points

kalabukdima committed:
Change HTTP endpoint to adjust for the current SDK

kalabukdima committed:
Set RUST_LOG in the network configs

eldargab commented on src/controller/stream.rs:
```diff
@@ -0,0 +1,341 @@
+    timeouts: Arc<TimeoutManager>,
+    // The first request is sent out to probe it's validity.
+    // Only after that multiple requests are sent in parallel.
+    trial: bool,
```
This strategy will slow down the streaming bootstrap, while saving just a little bit of money. Likely not worth it.

kalabukdima replied:
Okay, maybe you're right. I thought only the throughput is important, not the latency

Wiezzel commented on Cargo.toml:
```diff
@@ -1,37 +1,37 @@
 [package]
-name = "query-gateway"
-version = "1.0.8"
+name = "subsquid-query-gateway"
```
```suggestion
name = "sqd-query-gateway"
```

eldargab commented on src/types/request.rs:
```diff
@@ -0,0 +1,63 @@
+pub struct ClientRequest {
+    pub dataset_id: DatasetId,
+    pub query: ParsedQuery,
+    pub buffer_size: usize,
```
Everything below is definitely something not to be exposed even in the "secret mode".

kalabukdima replied:
While it's primarily designed for debugging, I don't see a reason to not allow users of their own gateway to set those parameters per request. They can set them in the config anyway.
But if we will host a public gateway, those should definitely be limited or removed at all.

eldargab replied:
>I don't see a reason to not allow users of their own gateway to set those parameters per request

It is not polite to expose to users low level details, that could change anytime. Also, I believe, those kinds of settings could be set once and for all.

kalabukdima replied:
Okay, I will hide it from public API, but I don't want to hardcode it because it slows down the development process if you have to recompile and restart the gateway (it's more than a minute) to test a new value

kalabukdima commented on src/controller/stream.rs:
```diff
@@ -0,0 +1,341 @@
+                            "Couldn't schedule query ({e:?}), retrying in {:?}",
+                            request.backoff
+                        );
+                        let backoff = request.backoff;
```
It's also useful when all the workers are unavailable for requests at the moment. Without a backoff you would just see "couldn't find worker for chunk {}" repeatedly and the stream would end because of a 404 error.

kalabukdima replied:
It may be a bit messy in the code, but there are two kinds of errors here:
- Errors returned immediately from `query_some_worker` function. I call those "query scheduling errors". They occur immediately, without `await`ing any async tasks. Therefore, retrying them immediately will probably cause the same result, so I've introduced a backoff.
- Errors received from the network, after `await`ing something. Those should probably be retried immediately after we get the error.

eldargab replied:
>So I think the usage of backoff doesn't really depend on which worker is queried.

That's not true. 

What is the most common cause of query failure according to our expectations? It should be a problem with a single worker or connectivity with a single worker. In such case a backoff timeout is just a waste of time.

>transport message queue is full, "query scheduling errors"

Sounds like something, that should not happen in the first place. But, if something like this happens, I'd say the best way to deal with that is to return the error immediately without any retries.

kalabukdima replied:
> In such case a backoff timeout is just a waste of time
Okay, agreed. But then a better approach would be to treat "not possible to send query" and "query failed" errors separately. I'll try to find a better solution. I don't really like how this function is implemented at the moment.

>Sounds like something, that should not happen in the first place
It can easily happen because we're scheduling multiple requests without an `await` point in between. So, for example, on a single-threaded runtime with buffer size > msg_queue_size it's guaranteed to happen. You're right, it's better to return such error and deal with it on a higher level.

kalabukdima replied:
I've handled the "not possible to send query" type of error separately. Now the buffer won't grow until the new queries can be successfully scheduled. It helps to avoid interrupting the stream if the `buffer_size` is too large for the current network conditions (many overloaded workers).

However, the transport queue (1000 messages) is filled up quite quickly and it probably becomes the bottleneck if multiple streams are run in parallel :(
@wiezzel is it safe to increase the outgoing messages buffer size?

kalabukdima commented on src/controller/task_manager.rs:
```diff
@@ -0,0 +1,79 @@
+
+use super::stream::StreamController;
+
+const MAX_PARALLEL_STREAMS: u8 = 5;
```
Yes, the main reason for this is to limit memory usage. If we want to host a public gateway, we surely don't want to make it unlimited

kalabukdima commented on src/controller/stream.rs:
```diff
@@ -0,0 +1,341 @@
+        let chunk = self.next_chunk.clone().expect("No next chunk to download");
+        let index: usize = self.next_index;
+        let future = self
+            .get_entire_chunk(index, chunk.clone())
```
I can propose the following solution.

When at least one chunk for the current stream doesn't fit in the response, immediately switch to the "slow" streaming mode. In this mode only a single chunk is requested simultaneously (no fetch ahead), but the partial results are sent to the client as soon as they are received. This would probably be enough for such a use case (which is not a recommended usage anyway), but would limit the memory usage to a single response size in this mode.

kalabukdima replied:
Okay, I've fixed it in a proper way. Each "slot" of the buffer now saves only one response. If the current first slot of the buffer is only a partial response, we send it to the client immediately and request the continuation.

kalabukdima commented on src/controller/stream.rs:
```diff
@@ -0,0 +1,341 @@
+                        let backoff = request.backoff;
+                        let send_query = send_query.clone();
+                        requests.spawn(async move {
+                            tokio::time::sleep(backoff).await;
```
Do you mean that backoff should be derived from request timeout? They are two separate things now.
- Request timeout limits the execution time for a single request.
- Backoff sets the delay between consecutive requests which immediately fail with an error, but this error is temporary (no workers available or transport message queue is full).

So even if all the requests are very fast, we don't want the backoff to be less than a certain limit to not overload already pressured components.

eldargab replied:
I mean, at first you time out request-response cycle, but then `request-response cycle + backoff time`.

kalabukdima replied:
I don't think so. The `start_time` is recorded at the `send_query` call, so it only limits the query execution time, without backoff

kalabukdima committed:
Remove trial requests

kalabukdima committed:
Handle partial responses with bounded memory usage

kalabukdima committed:
Make HTTP interface backward compatible

kalabukdima committed:
Rename crate to sqd-portal

eldargab commented on src/controller/stream.rs:
```diff
@@ -0,0 +1,358 @@
+        } else {
+            tracing::debug!("No more chunks available");
+        }
+        self.next_index += 1;
```
There is no need for this counter as `SlidingArray` does this already.

eldargab commented on src/controller/stream.rs:
```diff
@@ -0,0 +1,358 @@
+                Some(Ok(data))
+            }
+            Slot::Pending => unreachable!("First chunk is not ready"),
+        };
```
We popped the chunk, but haven't scheduled a new download. As a result we are downloading with less concurrency than requested.

eldargab commented on src/controller/stream.rs:
```diff
@@ -0,0 +1,358 @@
+            }
+        }
+
+        let index = self.buffer.first_index();
```
This value could be returned from `.push_front()` and `.first_index()` could be removed all together, to make idea behind `SlidingArray` more clear.

eldargab commented on src/network/priorities.rs:
```diff
@@ -0,0 +1,79 @@
+    priorities: HashMap<PeerId, Priority>,
+}
+
+impl WorkersPool {
```
My concern regarding the chosen method of evaluation of worker performance is that it is hard to reason about impact of the distant past behaviour.

For example, a worker could have a hiccup, but yet to receive quite a bit of queries at that moment (e.g. because it hosts some low replicated data chunks, that got a sudden demand). In such case, the worker will be significantly de-prioritised, but for how long? Might be for a very long time, depending on a usage pattern and overall network performance...

I suggest to track workers with a set of windowed (by time) counters and to compute resulting priorities at query time with some formula.

A set of counters could be:

* Number of "big" timeouts
* Number of times the worker failed to respond within a retry threshold
* Number of times worker responded within a 50% percentile
* Number of Busy errors
* Number of currently pending requests made to to this worker

kalabukdima replied:
These priorities are reset every epoch, which is 30mins currently, but I suspect it may get shorter later. This simple algo has already sped up the streaming about 2-4 times. 

I like what you're suggesting, but how about getting back to it after the initial release? It looks like improving performance without ability to run real world benchmarks

kalabukdima replied:
Done

kalabukdima committed:
Improve logs

- Report stream summary
- Add canonical log lines for HTTP requests
- Reorganize span fields
- Support JSON log format

kalabukdima commented on src/controller/timeouts.rs:
```diff
@@ -0,0 +1,95 @@
+        let num_infs = self.num_infs.load(Ordering::Relaxed);
+
+        // TODO: optimize time complexity
+        let kth = ((durations.len() + num_infs) as f32 * self.quantile).floor() as usize;
```
This algorithm ensures that the `timeout_percentile` of the slowest requests are retried on average. Its a generalization of what you proposed (wait until 90% of the requests complete, then retry the rest) but in the streaming manner - when the requests could start at different times. In this case I'm tracking the durations of completed requests and the current number of "ongoing" requests. The goal is to find a timeout which would "cut" the `1-timeout_quantile` fraction of the remaining requests.

I totally agree that this algorithm is hard to comprehend. But I couldn't fast a better one with the same performance and the same good qualities, namely:
- the requests are retried as soon as we know they fall into the given fraction of the slowest ones,
- if less than the `timeout_percentile` requests have finished so far, we will continue waiting for any of them to complete, 
- a request which has been just retried should be also timed out with the current duration

kalabukdima replied:
Fixed, almost as you suggested, but instead of using the longest existing time I always use the hard-coded timeout until there are 50 observations. Otherwise, some unexpected durations (either too fast or too long) could lead to unpredictable consequences.

kalabukdima committed:
Simplify timeout handling

Determine the request timeout at the moment of sending it.
Just return the default timeout if not enough data is available

kalabukdima committed:
Handle errors if the query couldn't be sent

- Remove backoff time — retry errors immediately or fail
- Don't extend the buffer if no more queries could be sent

kalabukdima commented on src/controller/timeouts.rs:
```diff
@@ -0,0 +1,95 @@
+            if let Some(timeout) = timeout {
+                tokio::select! {
+                    _ = tokio::time::sleep_until(start + timeout) => break,
+                    recv = current_timeout.changed() => {
```
Fixed. I've made it only get the timeout at the moment of sending a request, without adjusting it if it changes

kalabukdima committed:
Read chunks list from a shared file

Get rid of S3 API dependency

kalabukdima committed:
Make /query responses compatible with the archives

Convert JSON lines to a plain JSON array

kalabukdima committed:
Add more config options for the public portal

- Configure max number of parallel streams
- Configure max number of chunks for a single request

kalabukdima committed:
Add EVM and Substrate datasets

kalabukdima committed:
Rewrite StreamController in a polling fashion

kalabukdima committed:
Update to transport 1.0.25

kalabukdima committed:
Implement better priorities

kalabukdima committed:
Minor fixes

kalabukdima committed:
Improve logging

kalabukdima committed:
Add more prometheus metrics

kalabukdima committed:
Remove unnecessary Arcs
